### PR TITLE
[FEAT] Backend integration, directory restructure, failing test

### PR DESCRIPTION
**Pull request by indungu**
### What this PR does
Adds backend functionality to the web-app
Restructures the current working directory adding new files and folders, and copying existing files to new locations to achieve conventional app directory structure for Flask based single-module web apps

### Where to begin your review

Seeing as development focus is moving away from front end static UI design, the design of the site [here](https://indungu.gihub.io/yummy-recipes), shall remain the same for a well. Browsing through the site should give you a general feel of what the interaction with the dynamic webapp will feel like. The missing functionality shall also be added over the course of a few iterations

### How/What to test

Development of the backend component of this webapp follows test driven development (TDD) best practice and as such we take tests very seriously. As such, for best review experience install the following prerequisites
#### Prerequisites
1. *Python version*
    `$ python --version`
    `Python 3.6.2`
    If you have a lower version of Python please download the above version for your specific version 
    [here](https://www.python.org/downloads/release/python-362/) or alternatively a higher version 
    [here](https://www.python.org/downloads/), if one exists at the time of review.
2. *Virtual environment*
    `$ pip install virtualenv`
    alternatively
    `$ pip install pipenv`

#### Installation
**For MacOS or Linux**

```
  #!/bin/bash
  $ git clone -b dev https://github.com/indungu/yummy-recipes # Clone repo's dev branch
  $. cd yummy-recipes  # Switch to working directory
  $ virtualenv venv        # Install virtual envronment
  $ bin/activate             # Activate virtualenv
  $ pip install -r requirements.txt # Install package dependencies
  $ pip install pytest     # Install testing script
```
**For Windows**
In an elevated `cmd` prompt or PowerShell instance, enter the following commands

```
  #!c:\user\usr\
  > git clone -b dev https://github.com/indungu/yummy-recipes # Clone repo's dev branch
  > cd yummy-recipes  # Switch to working directory
  > virtualenv venv        # Install virtual environment
  > venv\Scripts\activate              # Activate virtualenv
  > pip install -r requirements.txt # Install package dependencies
  > pip install pytest                     # Install testing script
```
Once setup is complete type:

`python run.py`

Affirm that this starts the `development server` and gives the following output:

`* Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)`

While the server is running, navigate to the URL provided on a browser of choice and confirm that the page displays:
**Not Found**
`The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.`
 
Check the terminal on which the server was running for the following errors:

```
127.0.0.1 - - [18/Oct/2017 19:07:54] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [18/Oct/2017 19:07:55] "GET /favicon.ico HTTP/1.1" 404 
```
Press `ctrl+c` to stop the server and proceed to run the Unit test

`$ pytest` or `py.test` should run the tests in file `test_app.py` which is in the root of the working dir.
The following is the output to be expected:

```
============================= test session starts =============================
platform win32 -- Python 3.6.2, pytest-3.2.3, py-1.4.34, pluggy-0.4.0
rootdir: C:\Users\FAM\BOOTCAMP-XXIII\Challenges\Challenge-2\yummy-recipes, inifile:
collected 1 item

test_app.py F

============================= FAILURES ====================================
____________ HomePageTestCase.test_index_decorator_home_page_view _____________

self = <test_app.HomePageTestCase testMethod=test_index_decorator_home_page_view>

    def test_index_decorator_home_page_view(self):
        """Testing that the decorator renders the index.html template"""
        test_app = APP.test_client(self)
        response = test_app.get('/', content_type='html/text')
>       self.assertEqual(response.status_code, 200)
E       AssertionError: 404 != 200

test_app.py:17: AssertionError
========================= 1 failed in 3.02 seconds ==========================
```
This is the desired test result seeing as the path test to root yields a `HTTP status code` of `404` as did the browser when the request for `http://127.0.0.1:5000` was sent to the `dev server`.

### Notes
When this PR is merged it shall create the app directory structure and form  building block for future feature development whilst adhering to the TDD principles.

### Questions
Is there a better way to structure the entire source (repo) directory and especially the `app` subdirectory?

### CODE REVIEW

**Comment by ThubiMaina** [comment]
Good work I find it a good thing to commit your work in bits rather than having one commit. Version control should help you track changes

**Comment by indungu** [comment]
Duly noted with thanks @ThubiMaina I will try that too.

**Comment by indungu** [comment]
Commit `1fa3eb7` is inline with feedback received for a better directory structure.
current directory tree snippet is as shown below:

image

`app.py` was refactored to pass test cases in `test_app.py`

`test_app.py` was too refactored and an additional test added. The `reference import` was also changed.

However, in spite of the import refactor, running the `pytest` script generates the following error

```bat
============================= test session starts =============================
platform win32 -- Python 3.6.2, pytest-3.2.3, py-1.4.34, pluggy-0.4.0
rootdir: C:\Users\FAM\BOOTCAMP-XXIII\Challenges\Challenge-2\yummy-recipes, inifile:
collected 0 items / 1 errors

=================================== ERRORS ====================================
_____________________ ERROR collecting tests/test_app.py ______________________
ImportError while importing test module 'C:\Users\FAM\BOOTCAMP-XXIII\Challenges\Challenge-2\yummy-recipes\tests\test_app.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests\test_app.py:4: in <module>
    from app import APP
E   ModuleNotFoundError: No module named 'app'
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.35 seconds ===========================
```

**Comment by indungu** [comment]
Commit `1fa3eb7` also features changes to the `README.md`. 
These changes are:

* Markdown linting changes
* Content update

File: app/app.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+"""Main App module"""
+
+from flask import Flask, render_template, url_for
```
Comment (tinamorale) [inline app/app.py]: Remove unused imports

File: tests/test_app.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    def test_index_decorator_renders_home_page_view(self):
+        """Testing that the decorator renders the index.html template"""
+        test_app = APP.test_client(self)
+        response = test_app.get('/', content_type='html/text')
+        self.assertEqual(response.status_code, 200)
+
+    # They notice that the browser title reads Yummy Recipes | Welcome
+    def test_that_the_rendered_template_is_the_index_page(self):
+        """Ensure that the right template is rendered"""
+        test_app = APP.test_client(self)
```
Comment (tinamorale) [inline tests/test_app.py]: `test_app = APP.test_client(self)`This seems to appear in your methods. Use setup() for these. Programming rule *DRY* (Do not repeat yourself)

File: tests/test_app.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+class HomePageTestCase(unittest.TestCase):
+    """Home page unit test"""
+
+    # A user has heard about our amazing recipe app and decided to
+    # check it out
+    # They enter the url for the app in their browser and it
+    # displays a welcome screen
+    def test_index_decorator_renders_home_page_view(self):
+        """Testing that the decorator renders the index.html template"""
+        test_app = APP.test_client(self)
```
Comment (tinamorale) [inline tests/test_app.py]: `test_app = APP.test_client(self)`This seems to appear in your methods. Use setup() for these. Programming rule **DRY** (Do not repeat yourself)

File: app/app.py (hunk 1 above)
Comment (indungu) [reply app/app.py]: I ended up using them as shown in this commit `d57b997`
But I shall take note not to include unused imports.
Thanks.

File: tests/test_app.py (hunk 2 above)
Comment (indungu) [reply tests/test_app.py]: Thanks, I'm on it.

**Pull request merged**